base/sysinfo.jl

```diff
@@ -321,4 +323,81 @@ windows_version
 
 const WINDOWS_VISTA_VER = v"6.0"
 
+"""
+    Sys._can_execute(path::AbstractString)
+
+Returns `true` if the given `path` has executable permissions.
+"""
+function _can_execute(path::AbstractString)
+    # We use `access()` and `X_OK` to determine if a given path is executable
+    # by the current user.  `X_OK` comes from `unistd.h`.
+    X_OK = 1 << 0
+    @static if iswindows()
+        ccall(:_access, Cint, (Ptr{UInt8}, Cint), path, X_OK) == 0
+    else
+        ccall(:access, Cint, (Ptr{UInt8}, Cint), path, X_OK) == 0
+    end
+end
+
+"""
+    Sys.which(program_name::AbstractString)
+
+Given a program name, searches the current `PATH` to find the first binary with
+the proper executable permissions that can be run, and returns the absolute
+path. Raises `ArgumentError` if no such program is available.  If a relative or
+absolute path is passed in for `program_name`, that exact path is tested for
+executable permissions only, no searching of `PATH` is performed.
+"""
+function which(program_name::AbstractString)
+    # If prog has a slash, we know the user wants to determine whether the given
+    # file exists and is executable, and to not search the `PATH`.  Note that
+    # Windows can have either `\\` or `/` in its paths:
+    dirseps = @static if iswindows()
+        ["/", "\\"]
+    else
+        ["/"]
+    end
```

Comment:
Probably a tuple of chars here rather than an array of strings?

Reply:
Are tuples preferred here for performance considerations?